### PR TITLE
fix(settings,auth): Fix sumo login issue

### PR DESCRIPTION
**Pull request by dschom**
## Because

- Users of non-Sync OAuth RPs (e.g. SUMO) were receiving `verifyLoginCode` emails on signin even though the auth-server had correctly decided the session did not need verification (`mustVerify=false`). 
- We ultimately traced this back to `cachedSignIn` in fxa-settings firing `POST /session/resend_code` based solely on `sessionVerified=false` flag and ignoring the `mustVerify` flag. 
- The React rewrite dropped a `mustVerify` guard that fxa-content-server still had at `views/base.js`#`checkAuthorization`. 

## This pull request

- Moves the logic for sending the email into the front end. This should ensure that the code is only sent when the user is directed to a page requiring a code. 
- By moving the responsibility to the front end, we don't have to worry as much about the server side and client getting out of sync.

## Issue that this pull request solves

Closes: FXA-12972

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).
- [x] I have manually reviewed all AI generated code.

## How to review (Optional)

- Key files/areas to focus on:
  - `packages/fxa-settings/src/pages/Signin/utils.ts` — the front end fix (one new `&& mustVerify` guard).
  - `packages/fxa-auth-server/lib/routes/session.js` — the new field on `/session/status`.
- Suggested review order: auth-server route change → auth-client type → fxa-settings gate → unit tests → remote tests.
- Risky/complex parts: no end-to-end functional smoke test was added. See comments below...

## Other information

Tracking this bug down was a little twisty, so a full diagnosis of problem is attached to FXA-12972

We also tried to create a smoke-test to cover this edge case but could not, since the state hinged on the `signinConfirmation.skipForNewAccounts` option which is set to 48 hours currently. Because this time lapse is quite large, there's no really a good way to trip this state in a functional test. The interactions between these flags and options (mustVerify, sessionVerified, accountVerified, lockedAt, skipForNewAccounts, etc...) is pretty complex, so it'd be worth while clearly demarcating these states. 

The unit-level coverage proves the fix; however, in order to get functional test coverage we need either a provisioned long-lived fixture account on each environment or an admin-API path to alter an existing account state. A follow spike was added to look into how to make get better test coverage around this.

Also worth noting that based on feedback, we decided not expose the mustVerify flag to the front end, and instead pivoted to moving the send logic entirely to the front end.

### CODE REVIEW

File: packages/fxa-settings/src/pages/Signin/utils.ts [hunk 1]
```diff
@@ -161,9 +162,12 @@ export const cachedSignIn = async (
       verificationReason = VerificationReasons.SIGN_UP;
       verificationMethod = VerificationMethods.EMAIL_OTP;
       !isOauthPromptNone && (await session.sendVerificationCode());
-    } else if (!sessionVerified) {
-      // we are inferring verification method and verification reason here
-      // ideally we would check the server directly to allow for more verification reasons
+    } else if (!sessionVerified && mustVerify) {
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/utils.ts]: @LZoog All the other changes, support this small update to the front end logic... I'm still pondering if there's a simpler way to achieve this, but basically we should invoking line 176, if the session doesn't actually require verification...

Comment (vpomerleau) [reply packages/fxa-settings/src/pages/Signin/utils.ts]: Smells a bit to me that we are handling these checks client-side. Perhaps we could update `/session/resend_code` to confirm whether the session still needs verification and return if a code was sent or not, to inform client-side navigation?

Comment (dschom) [reply packages/fxa-settings/src/pages/Signin/utils.ts]: I also considered this and I mostly agree with what you are saying. However, this was deemed a riskier option and it still doesn't entirely clean this up... 

The two issues with that approach seem to be:

1. In this case, we would make the call, and then fail. The reality is that we do not need the extra call. So the client side is still do something unnecessary / wrong.

2. The other, probably bigger, concern is that this could lead to scenarios where a user is being prompted by the UI for a code, the do something to invoke `session.resendCode`. We then start erroring on the server side, and the user is now stuck. 

I guess we've had enough issues around verified session states, that I'm a little hesitant to do something that might result in other issue. In someways, I think exposing the account model state to the front end more fully is better than making unnecessary calls and then seeing waiting for errors. It's possible that by exposing this flag to the front end, we can actually make better decisions about what to do. This scenario is actually a good example of that.

The other thing I don't quite get, and maybe this yet another way to address the problem. Why are we invoking `resendCode` here? Shouldn't that send happen on the page that is asking for the code?

Comment (vpomerleau) [reply packages/fxa-settings/src/pages/Signin/utils.ts]: The server generally handles sending the code. It's not usually triggered by the frontend except here. There must be something else funky that's leading to the exception.

Comment (dschom) [reply packages/fxa-settings/src/pages/Signin/utils.ts]: @vpomerleau I'm moving this back to draft and will iterate a little more on it. I'm going to try and shift the code that sends the server code request closer to the navigation. That will keep the a bit more tightly coupled and hopefully prevent misfires.

Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/utils.ts]: Thanks for looking at this and then looking again, Dan. Fwiw so long as we call this out in https://mozilla-hub.atlassian.net/browse/FXA-13180 as something we may want to reverse I'm OK with the first solution because we need to investigate a lot more there but I appreciate you looking at it again since we have gotten away with not needing `mustVerify` in the FE until now.

File: packages/fxa-settings/src/pages/Signin/utils.ts [hunk 2]
```diff
@@ -295,6 +292,16 @@ export async function handleNavigation(navigationOptions: NavigationOptions) {
       sendFxaLogin(navigationOptions);
     }
 
+    // If we are about to direct a user to /signin_totp_code, where they will be prompoted for a
+    // totp code and we know their session isn't fully verified then send them an otp code.
+    if (
```
Comment (dschom) [inline packages/fxa-settings/src/pages/Signin/utils.ts:300]: @vpomerleau I threw out the previous approach, and switched to this.

File: packages/fxa-settings/src/pages/Signin/utils.ts [hunk 3]
```diff
@@ -166,16 +159,20 @@ export const cachedSignIn = async (
     if (!emailVerified) {
       verificationReason = VerificationReasons.SIGN_UP;
       verificationMethod = VerificationMethods.EMAIL_OTP;
-      !isOauthPromptNone && (await session.sendVerificationCode());
+      // We should send the code prior to nav. That couples with an actual navigation action, where it belongs.
+      // !isOauthPromptNone && (await session.sendVerificationCode());
```
Comment (LZoog) [inline packages/fxa-settings/src/pages/Signin/utils.ts:164]: Stray old code comment left in!

I think I'm a little confused by the `prompt=none` changes, or how it would've impacted the SUMO unverified session state, which is not `prompt=none`. We _want_ it to reject, (RP decides what to do and takes the user back to login), send an email verification code, and take the user to `confirm_signup_code`. I'll test locally though.

File: packages/fxa-settings/src/pages/Signin/utils.ts [hunk 4]
```diff
@@ -166,16 +159,20 @@ export const cachedSignIn = async (
...
     } else if (!sessionVerified) {
       // we are inferring verification method and verification reason here
       // ideally we would check the server directly to allow for more verification reasons
       verificationReason = VerificationReasons.SIGN_IN;
       if (totpIsActive) {
         verificationMethod = VerificationMethods.TOTP_2FA;
       } else {
+        // Note when using this method, the expectation is that we will send the user
+        // a verification code, if and only if the navigation handler directs them to
+        // a verification code page. There are some RPs that require email_otp verification,
```
Comment (LZoog) [inline packages/fxa-settings/src/pages/Signin/utils.ts:174]: >There are some RPs that require email_otp verification so for these RPs, we can just pass the user straight through.

Should this be, "... some RPs that require email_otp verification [kept in an auth-server config] so for _other_ RPs we can pass the user straight through"?

File: packages/fxa-settings/src/pages/Signin/utils.ts [hunk 5]
```diff
@@ -295,6 +292,23 @@ export async function handleNavigation(navigationOptions: NavigationOptions) {
       sendFxaLogin(navigationOptions);
     }
 
+    // If we are about to direct the user to an email-OTP code entry page
+    // (signin_token_code for an unverified session, or confirm_signup_code for
+    // an unverified email) and their session isn't fully verified, send them a
+    // code now. The send is coupled to the navigation so RPs that pass the user
+    // straight through (e.g. prompt=none) don't trigger an extraneous email.
```
Comment (LZoog) [inline packages/fxa-settings/src/pages/Signin/utils.ts]: > or confirm_signup_code for an unverified email) and their session isn't fully verified, send them a code now.

Just a nit but this reads a little funky to me, since an unverified email/account will always have an unverified session.

File: packages/fxa-settings/src/pages/Signin/utils.ts (hunk 4 above)
Comment (dschom) [reply packages/fxa-settings/src/pages/Signin/utils.ts:174]: I'll clean up the wording though cause it's not explicit enough. I think the 'we can pass the user straight through', is just too ambiguous, and we are interpreting that differently. I meant we can pass the user straight through to the navigation handler, not that we can have them skip verification or anything like that.

What this is really saying is that, we should let the navigation handler decide whether or not to send the otp email. Ultimately, the navigation handler holds the ground truth for whether or not the user will be directed to an OTP verification page, and this is the best point in time to decide whether or not to send an email."

File: packages/fxa-settings/src/pages/Signin/utils.ts (hunk 3 above)
Comment (LZoog) [reply packages/fxa-settings/src/pages/Signin/utils.ts:164]: Just following up here, not sure if this was related to the bug BUT this all seems to work as expected. We reject with an error back to the RP on an unverified account, and 123done reroutes back without `prompt=none`, and I'm getting the email code when I expect. 👍

**Review by LZoog — APPROVED** [review]
Lots of Slack conversations about this - we'll have a clean up issue filed to look at this more holistically. This is reproducing in main and in this branch as well:

> I'm also getting a token code email going through the RP redirect with that email, but signin_token_code page is correctly skipped
